### PR TITLE
pt:Removed the data system implemented in PyTorch

### PR DESCRIPTION
**Pull request by CaRoLZhangxy**
The old code has been retained for reference by @Chengqian-Zhang for de-noise implementation.

### CODE REVIEW

File: deepmd/utils/data.py [hunk 1]
```diff
@@ -129,6 +129,15 @@ def __init__(
...
         self.modifier = modifier
+        # calculate prefix sum for get_item method
+        self.nframes = 0
+        i = 1
+        self.prefix_sum = [0] * (len(self.dirs) + 1)
+        for item in self.dirs:
+            frames = self._get_nframes(item)
+            self.prefix_sum[i] = self.prefix_sum[i - 1] + frames
+            i += 1
+            self.nframes += frames
```
Comment (wanghan-iapcm) [inline deepmd/utils/data.py:140]: what is `prefix_sum` ? explanation in doc is needed.

File: deepmd/utils/data.py [hunk 2]
```diff
@@ -440,6 +468,30 @@ def _shuffle_data(self, data):
...
+        if self.data_dict["coord"]["high_prec"]:
+            coord = path.load_numpy().astype(GLOBAL_ENER_FLOAT_PRECISION)
+        else:
+            coord = path.load_numpy().astype(GLOBAL_NP_FLOAT_PRECISION)
+        if coord.ndim == 1:
+            coord = coord.reshape([1, -1])
+        nframes = coord.shape[0]
+        return nframes
+
+    def preprocess(self, data):
```
Comment (wanghan-iapcm) [inline deepmd/utils/data.py:485]: doc string needed. 
is the preprocess used only for the torch code? 
suggest to rename the function in a more self-explanatory way.

File: deepmd/utils/data.py [hunk 3]
```diff
@@ -129,6 +129,16 @@ def __init__(
...
+        # calculate prefix sum for get_item method
+        self.nframes = 0
+        i = 1
+        # The prefix sum stores the range of indices contained in each directory, which is needed by get_item method
+        self.prefix_sum = [0] * (len(self.dirs) + 1)
+        for item in self.dirs:
+            frames = self._get_nframes(item)
+            self.prefix_sum[i] = self.prefix_sum[i - 1] + frames
+            i += 1
+            self.nframes += frames
```
Comment (njzjz) [inline deepmd/utils/data.py:141]: It looks you just rewrote np.cumsum and np.sum

Comment (caic99) [reply deepmd/utils/data.py:141]:
```suggestion
        # calculate prefix sum for get_item method
        frames_list = [self._get_nframes(item) for item in self.dirs]
        self.nframes = np.sum(frames_list)
        # The prefix sum stores the range of indices contained in each directory, which is needed by get_item method
        self.nframes = np.cumsum([0] + frames_list).tolist()
```

@CaRoLZhangxy Plz try if this is OK

Comment (njzjz) [reply deepmd/utils/data.py:141]: Or use `itertools.accumulate` and `sum` without numpy

File: deepmd/pt/utils/dataset.py [hunk 4]
```diff
@@ -575,44 +24,22 @@ def __init__(
...
+        self._data_system = DeepmdData(
+            sys_path=system, shuffle_test=shuffle, type_map=self._type_map
         )
+        self._data_system.add("energy", 1, atomic=False, must=False, high_prec=True)
+        self._data_system.add("force", 3, atomic=True, must=False, high_prec=False)
+        self._data_system.add("virial", 9, atomic=False, must=False, high_prec=False)
         self.mixed_type = self._data_system.mixed_type
         self._ntypes = self._data_system.get_ntypes()
-        self._natoms = self._data_system._natoms
+        self._natoms = self._data_system.natoms
```
Comment (njzjz) [inline deepmd/pt/utils/dataset.py:35]: The public method is preferred
```suggestion
        self._natoms = self._data_system.get_natoms()
```

File: deepmd/pt/utils/dataset.py [hunk 5]
```diff
@@ -575,44 +24,22 @@ def __init__(
...
-            type_map=self._type_map,
-            type_split=type_split,
-            noise_settings=noise_settings,
-            shuffle=shuffle,
+        self._data_system = DeepmdData(
+            sys_path=system, shuffle_test=shuffle, type_map=self._type_map
         )
+        self._data_system.add("energy", 1, atomic=False, must=False, high_prec=True)
+        self._data_system.add("force", 3, atomic=True, must=False, high_prec=False)
+        self._data_system.add("virial", 9, atomic=False, must=False, high_prec=False)
```
Comment (njzjz) [inline deepmd/pt/utils/dataset.py:32]: I don't quite understand why these are hard-coded here - what will happen if we fit the dipole other than energy?

File: deepmd/utils/data.py [hunk 6]
```diff
@@ -249,6 +254,25 @@ def check_test_size(self, test_size):
         else:
             return None
 
+    def get_item(self, index: int) -> dict:
+        """Get a single frame data . The frame is picked from the data system by index.
```
Comment (wanghan-iapcm) [inline deepmd/utils/data.py:258]:
```suggestion
        """Get a single frame data . The frame is picked from the data system by index. The index is coded across all the sets.
```

File: deepmd/utils/data.py [hunk 7]
```diff
@@ -249,6 +254,25 @@ def check_test_size(self, test_size):
...
+        Parameters
+        ----------
+        index
+            index of the frame
+        """
+        for i in range(
+            0, len(self.dirs)
+        ):  # note: if different sets can be merged, prefix sum is unused to calculate
+            if index < self.prefix_sum[i]:
+                break
```
Comment (njzjz) [inline deepmd/utils/data.py:269]: I guess you need [bisect](https://docs.python.org/3/library/bisect.html). This can be simply written as
```suggestion
        i = bisect.bisect_right(self.prefix_sum, index)
```

File: deepmd/utils/data.py [hunk 8]
```diff
@@ -249,6 +254,25 @@ def check_test_size(self, test_size):
         else:
             return None
 
+    def get_item(self, index: int) -> dict:
```
Comment (njzjz) [inline deepmd/utils/data.py:257]: It seems this method can only be used by pt (since it calls `reformat_data_torch`).

File: deepmd/pt/utils/dataset.py (hunk 5 above)
Comment (njzjz) [reply deepmd/pt/utils/dataset.py:32]: @iProzd do you know what the design is here?

**Review by wanghan-iapcm — APPROVED** [review]

**Review by njzjz — APPROVED** [review]

**Review by Chengqian-Zhang — APPROVED** [review]

**Review by caic99 — APPROVED** [review]

File: deepmd/pt/utils/dataset.py (hunk 5 above)
Comment (iProzd) [reply deepmd/pt/utils/dataset.py:32]: This should be placed into loss or model requirements indeed. The hard-coded here was then not well-designed.
Btw in tf it's a global variable called `data_requirement`, which is not flexible way for multi-task training when different models have different requirements. 

I think I will fix this part in training code cleaning PR.

**Review by iProzd — APPROVED** [review]
